debug stream

h-2 opened the pull request.

h-2 commented on include/seqan3/core/debug_stream.hpp:
```diff
@@ -0,0 +1,202 @@
+template <alphabet_concept alphabet_t>
+inline debug_stream_t & operator<<(debug_stream_t & s, alphabet_t const l)
+//!\cond
+    requires !(requires (alphabet_t l) { { std::cerr << l }; })
```
any ostream, but cerr is ok

h-2 commented on include/seqan3/core/debug_stream.hpp:
```diff
@@ -0,0 +1,202 @@
+ * just as if it were a string, i.e. `std::vector<dna4>{dna4::C, dna4:G, dna4::A}` is printed as "CGA".
+ *
+ * In all other cases the elements are comma separated and the range is enclosed in brackets, i.e.
+ * `std::vector<int>{3, 1, 33, 7}` is printed as "[3,1,33,7]".
```
"the range library" will only be a dependency before c++20 and people will never really know that we use it so there is no benefit in referring to it, I think.

h-2 commented on test/snippet/alphabet/structure/dot_bracket3.cpp:
```diff
@@ -10,7 +10,7 @@ std::vector<dot_bracket3> vec{dot_bracket3::UNPAIRED, dot_bracket3::PAIR_CLOSE,
 vec[1] = dot_bracket3::PAIR_OPEN;
 for (dot_bracket3 chr : vec)
-    std::cout << chr;  // .()
+    std::cout << to_char(chr);  // .()
```
someone will have to do a PR and update all the snippets and example, but it is out of scope of this PR

h-2 commented on test/snippet/core/debug_stream.cpp:
```diff
@@ -0,0 +1,53 @@
+std::vector<dna5> vec{"ACGT"_dna5};
+//std::cout << vec;
+// but all types that model std::ranges::InputRange are printable to the debug_stream:
+debug_stream << vec;            // prints "[A,C,G,T]"
```
good catch, was changed quite late in the PR

h-2 commented:
> debug_stream is a little long compared to cout, no?

it's 3 longer than `std::cout` :wink: 

> This will be used extensively so maybe here shortness comes before expressiveness...
how about dout 

we thought about this (`dout` was a candidate), but couldn't agree on anything other than `debug_stream`.

h-2 commented on include/seqan3/core/debug_stream.hpp:
```diff
@@ -0,0 +1,202 @@
+template <std::ranges::InputRange rng_t>
+inline debug_stream_t & operator<<(debug_stream_t & s, rng_t && r)
+//!\cond
+    requires requires (reference_t<rng_t> l) { { debug_stream << l }; }
```
> does the range overload not need 

no, we want to overload this also for view, because views of `char` are now printed in string-notation.

> maybe try to print a view in a test.

it works, I added an example

h-2 commented:
I tihnk I addressed everything...

joergi-w commented on include/seqan3/core/debug_stream.hpp:
```diff
@@ -0,0 +1,219 @@
+    std::ostream *s = nullptr;
+
+public:
+    /*!\name Constructor, destructor and assignment.
```
Constructors (they are multiple)

joergi-w commented on include/seqan3/core/debug_stream.hpp:
```diff
@@ -0,0 +1,219 @@
+{
+private:
+    //!\brief Pointer to the output stream.
+    std::ostream *s = nullptr;
```
Please give class members a proper name, like `stream`

joergi-w commented on include/seqan3/core/debug_stream.hpp:
```diff
@@ -0,0 +1,219 @@
+    {}
+    //!\}
+
+    /*!\name Miscellenous
```
typo: miscellaneous

joergi-w commented on include/seqan3/core/debug_stream.hpp:
```diff
@@ -0,0 +1,219 @@
+ * \snippet test/snippet/core/debug_stream.cpp usage
+ *
+ * The actual underlying stream that is printed to defaults to std::cerr, but can be changed via
+ * seqan3::debug_stream_type::set_stream. You can set any kind of output stream, e.g. a std::ostringstream or a
```
...set_stream`()` 
Add parentheses to show that it is a function.

h-2 commented:
I have pushed changes addressing most of your comments and I have also added flags. I have added support for std::ios_base::fmtflags and also seqan3::fmtflags2 which contain a utf8 option that could be useful for @marehr 's stuff (it isn't used, yet, but I implemented another useful option as proof that it works).
